examples/gno.land/p/gnoland/commondao/proposal.gno

```diff
@@ -0,0 +1,174 @@
+package commondao
+
+import (
+	"errors"
+	"std"
+	"time"
+)
+
+// DefaultQuorum defines the default quorum required to tally proposal votes.
+const DefaultQuorum = 0.34 // 34%
+
+const (
+	StatusActive   ProposalStatus = "active"
+	StatusFailed                  = "failed"
+	StatusExecuted                = "executed"
+)
+
+const (
+	ChoiceAbstain VoteChoice = ""
+	ChoiceYes                = "YES"
+	ChoiceNo                 = "NO"
+)
+
+var (
+	ErrInvalidCreatorAddress      = errors.New("invalid proposal creator address")
+	ErrProposalDefinitionRequired = errors.New("proposal definition is required")
+	ErrStatusIsNotActive          = errors.New("proposal status is not active")
+)
+
+type (
+	// ProposalStatus defines a type for different proposal states.
+	ProposalStatus string
+
+	// VoteChoice defines a type for proposal vote choices.
+	VoteChoice string
+
+	// Proposal defines a DAO proposal.
+	Proposal struct {
+		id             uint64
+		status         ProposalStatus
+		definition     ProposalDefinition
+		creator        std.Address
+		record         *VotingRecord
+		statusReason   string
+		votingDeadline time.Time
+		createdAt      time.Time
+	}
+
+	// ProposalDefinition defines an interface for custom proposal definitions.
+	// These definitions define proposal content and behavior, they esentially
+	// allow the definition for different proposal types.
+	ProposalDefinition interface {
+		// Title returns the proposal title.
+		Title() string
+
+		// Body returns the proposal body.
+		// It usually contains the proposal description and other elements like proposal parameters.
+		Body() string
+
+		// VotingPeriod returns the period where votes are allowed after proposal creation.
+		// No more votes should be allowed once this period is met. It is used to calculate
+		// the voting deadline from the proposal's creationd date.
+		VotingPeriod() time.Duration
+
+		// Quorum returns the percentage of members that must vote to be able to pass a proposal.
+		// This is an optional value. DAOs use a default value when quorum is zero or invalid.
+		// Its value must be between 0 and 1, being 1 = 100% of member votes.
+		Quorum() float64
+
+		// Tally counts the number of votes and verifies that there is consensus.
+		// Tally fails when none of the vote choices wins over the others.
+		Tally(r ReadOnlyVotingRecord, memberCount int) (success bool)
+	}
+
+	// Validable defines an interface for proposal definitions that require state validation.
+	// Validation is done before execution and normally also during proposal rendering.
+	Validable interface {
+		// Validate validates that the proposal is valid for the current state.
+		Validate() error
+	}
+
+	// Executable defines an interface for proposal definitions that modify state on approval.
+	// Once proposals are executed they are archived and considered finished.
+	Executable interface {
+		// Execute executes the proposal.
+		Execute() error
+	}
```

Comment:
Why is this defined as an isolated interface? What can a proposal do apart from executing some code when approved?

Reply:
The interface is separate because not all proposal types would be executable, meaning that they would modify state. Some proposal types, like for example a general (text) proposals or budget approval proposals don't modify it.

Having it as a separate interface makes the proposal definition interface smaller and doesn't enforce to have empty `Execute()` methods on proposals that won't modify the state though a callback.

So, non executable proposals would highly likely only change their status when finished or "executed", changing only proposal state. The status change is something that happens independently of the `Execute()` method, status depends only on votes and deadline, there is no execution of a callback in these cases.